### ISSUE TITLE
How to register duplicate handlers

### ISSUE THREAD

**Issue by Condra963** [jbogard/MediatR.Extensions.Microsoft.DependencyInjection#47]
Hi,

I have created a pull request in MediatR and MediatR.DI to handle registering duplicate IRequestHandlers.
jbogard/MediatR#320 and #46 

To summarize the discussions:
You have for example a **Package** that you use that already contains **MediatR Handling**. Now you are creating your own **Software** depending on that **Package** and you want to overwrite these handlers because you want to modify this handling to more software specific handling.
At this moment, **MediatR.DI** just register the 2 handlers that are exactly the same like twins, only the implementation is different. So when you will call handlers request, it will call one of the 2 and you are not sure which one.

The change i made in these pull requests give you the opportunity to say which Handler can be overwritten. It was a solution, but not necessary the most satisfying one. For this case Jimmy asked me to open an Issue so we can discuss about it. What are the possibilities to implement this?

Thanks!

**Comment by lilasquared** [comment]
Some Ideas discussed in the previous thread summarized (plus my own)

1. (new idea) Have smaller packages that you depend on.  If you are truly replacing an entire suite of functionality then would it be possible to have that suite of functionality in its own package, that this software instance would just not reference?  Then there are not two instances of the same handler interface anymore.
2. Instead of a second `IRequestHandler` introduce an `IPipelineBehavior`.  Below is an example.
```csharp
// This exists in the *Package* you are referring to
public class Ping : IRequest<Pong> { }
public class PingHandler : IRequestHandler<Ping, Pong> 
{
   ... 
}
...
 // This exists in the *Software* you are referring to
public class PingBehavior : IPipelineBehavior<Ping, Pong>
{
   public async Task<Pong> Handle(Ping request, CancellationToken cancellationToken, RequestHandlerDelegate<Pong> next)
   {
        // implement overridden handle logic here ignoring next, which would execute the handler from the package
   }
}
```
3. Extend the behavior of AddMediatR to allow an option for overwriting in some way based on some condition.
```
services.AddMediatR().WithOverridesOf(x => Attribute.IsDefined(x, typeof(OverwriteableAttribute)));
```

**Comment by Condra963** [comment]
1. Actually not the solution and even if we lower the replacing behavior we still have to do some so the feature will still be useful.
2. If i understand PipelineBehavior correctly is this not how we should use it i think.
3. It is better to write one condition how you will override twin handlers because if you work in a team of 20 people it is possible they will write 20 different conditions. It is better to define 1 use case that does not make a lot of difference for anyone. 

Make it configurable like Jimmy says is maybe some solution. But will we make it configurable for each handler?

What a suggestion could be is an exception should be in there place if you register 2 times the same `IRequestHandler`, because that's only allowed for `INotificationHandler`. Then the user is already notified _**ow ok, i need to define which handler i want to use**_.
And above that i suggest configuration like:
```
{
  "MediatR": {
    "Requests": {
      "Request1": "OverwrittenRequest1Handler"
    }
  }
}
```
And if they register twin `IRequestHandlers` the user should be forced to configure this. 
Next to that we could also create something that you can pass this by an extensions method if you want to configure this in code instead of `appsettings`.

Is this some behavior that would be agreeable?

**Comment by Condra963** [comment]
FYI: `IPipelineBehavior` is not also automatically registered so that would not do anything about the extra work i have to do. 

Is there a reason `IPipelineBehavior` is not automatically registered?

**Comment by jbogard** [comment]
Configuration would need to be only in code, and only assembly/type based.

The easiest thing would be to enforce assembly order/preference.

**Comment by jbogard** [comment]
I also wouldn’t want handled cherry-picking since the existing DI API
already easily supports that. I only want to facilitate assembly-level
decisions.

**Comment by Condra963** [comment]
`.AddMediatR(Assembly1, Assembly2, Assembly3, ...)`
Assembly1 should be all the basic handlers, Assembly2 will overwrite Assembly1's basic handlers that are equal to Assembly2, Assembly3 will do the same for Assembly1 and Assembly2, and so on.

Something like that?

**Comment by jbogard** [comment]
Yep. The next question - should this be the default behavior? Originally I didn't want to allow it because people accidentally registered multiple handlers, not realizing only one would get called. Maybe an enum on registration behavior?

**Comment by lilasquared** [comment]
@jbogard what about a chain instead of an enum?

```
services.AddMediatR(Assembly1, Assembly2);

services.AddMediatR(Assembly1, Assembly2).WithOverrides(Assembly3);
```

**Comment by Condra963** [comment]
@jbogard can i ask why you didn't in the first place throw an `Exception` when they register multiple `IRequestHandlers` that are equal? Because it's behavior is 1 on 1, isn't it? Then people will be noticed they register multiple twin `IRequestHandlers`.

@lilasquared I can find myself into the chain. But do we need to think about the possibility that multiple assemblies can be a mix of what we would override or ignore?

The question is how far you want to go with this feature i think.
```
// Default behavior, in this case it will depend on the order of the assemblies. Scenarios where this will turn out badly? Possibility to add an ignore as well here?
services.AddMediatR(Assembly1, Assembly2, Assembly3);
// Suggested chain, this case will only takes the overrides of Assembly3. Here you have the choice to add Assembly2 or not to the override behavior.
services.AddMediatR(Assembly1, Assembly2).AddMediatRWithOverrides(Assembly3);
// Extension on the suggestion before, Only this type will overwrite his twin it will not look for any other overrides.
services.AddMediatR(Assembly1, Assembly2).AddMediatRWithOverrides(OverwrittenRequestHandlerType);
// Extension on suggestion before, OverwrittenRequestHandlerType will be ignored when all other types of Assembly3 will be registered.
services.AddMediatR(Assembly1, Assembly2).AddMediatRWithOverrides(Assembly3, OverwrittenRequestHandlerType);
```

This is nothing concrete, it's only to broaden our perspective a little bit.

**Comment by jbogard** [comment]
@Condra963 I used to throw an exception for multiple handlers inside of the base MediatR library, but I took it out and can't remember why.

**Comment by jbogard** [comment]
Chaining for configuration is tough to work with. If you want configuration, it's better to have that as something you pass in:

```c#
services.AddMediatR(cfg => cfg.AllowRegistrationOverrides, Assembly1, Assembly2, Assembly3);
```

**Comment by Condra963** [comment]
@jbogard The current behavior of `AddMediatR` is that it add both to the `ServiceCollection`. I investigated into the `ServiceCollection` of `Microsofts DependencyInjection` why it's allowed and you can register multiple times `ITransportation` and call `GetServices<ITransportation>` and you will get a collection of cars, motorcycles, trucks,... So that is why containers will not throw an `Exception` when you register multiple `IRequestHandlers`. But `AddMediatR` should throw an `Exception` i think because with `IRequestHandlers` you require 1 on 1 behavior. But if you once removed there could be a possible reason but i cannot figure out either which reason. Sorry.

 For configuring the overrides, can we override in order of the assemblies like we discussed before and pass configuration for the handlers we want to handle otherwise?

**Comment by jbogard** [comment]
I don't want configuration for individual handlers - that's too complicated. I assume you have multiple assemblies, and an assumption that later handlers will override previous ones.

**Comment by jbogard** [comment]
Check out #53 where you can now configure registration options. That's a perfect place now to add a method to allow overriding/ordering of assemblies.

**Comment by jbogard** [comment]
Also, would just having a way to just the duplicates overridden work? I don't want to get into a handler-by-handler basis. It seems like you'd really just want assembly B to override handlers in assembly A.

**Comment by Condra963** [comment]
Ok, i will look into #53 and implement something. If i'm done i will reference this issue to the pull-request and we can see from there if it's a acceptable solution.

**Comment by ben-foster-cko** [comment]
I'm looking for similar behaviour for a different reason. When running some of our integration tests I want to swap out the implementation of a particular handler. The issue is that Mediatr does not seem to honour the registration, even if done before calling `AddMediatR`, for example:

```
            if (_configuration.GetValue("Payments:Queuing:ProcessSynchronously", false))
            {
                // Overrides the default SQS handler with an in-process one
                services.AddTransient(typeof(IRequest<Enqueue>), typeof(InProcessEnqueueHandler));
            }

            services.AddTransient(typeof(IPipelineBehavior<,>), typeof(LoggingMediatorBehavior<,>));
            services.AddTransient(typeof(IPipelineBehavior<,>), typeof(SequentialProcessingBehavior<,>));
            services.AddMediatR(typeof(PaymentsRegistry).Assembly);
```
